### PR TITLE
[IMP] models: use new records only in onchange

### PR DESCRIPTION
**Pull request by rco-odoo**
Using new records in onchange() simplifies code: the environment flag `in_onchange` is gone, and all special cases correspond to new records. The expression `record._origin` is now available on *all* records, and returns the *actual* record corresponding to `record`.

### CODE REVIEW

File: odoo/fields.py [hunk 1]
```diff
@@ -2267,27 +2278,32 @@ def process(ids):
         elif isinstance(value, (list, tuple)):
             # value is a list/tuple of commands, dicts or record ids
             comodel = record.env[self.comodel_name]
+            if record and not record.id:
+                # record is new: the value of self is new records
+                browse = lambda id_: comodel.new(origin=comodel.browse(id_))
+            else:
+                browse = comodel.browse
             # determine the value ids; by convention empty on new records
             ids = OrderedSet(record[self.name].ids if record.id else ())
```
Comment (mreficent) [inline odoo/fields.py]: Could you check for this line the proposal I made in https://github.com/odoo/odoo/pull/24458? thanks

File: addons/survey/wizard/survey_invite.py [hunk 2]
```diff
@@ -128,7 +128,7 @@ def _onchange_partner_ids(self):
             if not self.survey_id.users_can_signup:
                 invalid_partners = self.env['res.partner'].search([
                     ('user_ids', '=', False),
-                    ('id', 'in', self.partner_ids.ids)
+                    ('id', 'in', self.partner_ids._origin.ids)
```
Comment (rco-odoo) [inline addons/survey/wizard/survey_invite.py]: Let `a` denote an actual record, `b` denote a new record with origin `b1`, and `c` denote a new record without origin.  Then we have:
```python
(a + b + c)._origin = a + b1
(a + b + c).ids = [a.id, b1.id]
```
In other words, `recs.ids == recs._origin.ids`.

File: odoo/addons/test_new_api/tests/test_new_fields.py [hunk 3]
```diff
@@ -929,24 +929,106 @@ def test_31_prefetch(self):
...
+        self.assertEqual(new_msg0.body, "XXX")
+        self.assertEqual(new_msg1.body, "YYY")
+        self.assertEqual(new_msg2.body, "ZZZ")
+
+        # original records have not been modified
+        self.assertFalse(msg0.discussion)
+        self.assertFalse(msg1.discussion)
+
+        self.assertEqual(msg0.body, "XXX")
+        self.assertEqual(msg1.body, "WWW")
```
Comment (rco-odoo) [inline odoo/addons/test_new_api/tests/test_new_fields.py]: Add tests for accessing all kinds of relational fields (many2one, one2many, many2many) with all special cases.

File: odoo/fields.py [hunk 4]
```diff
@@ -1119,12 +1125,22 @@ def determine_value(self, record):
...
-            fields = record._field_computed[self]
-            with record.env.protecting(fields, record):
-                self._compute_value(record)
+            self.compute_value(record)
+        elif record._origin:
+            value = record._origin[self.name]
+            if self.type == 'many2one' and self.delegate and value:
+                # special case: the parent record is a new record, too
+                value = value.new(origin=value)
+            value = self.convert_to_cache(value, record)
```
Comment (rco-odoo) [inline odoo/fields.py]: Should this convert actual records to new records if `self` is an x2many field?

File: odoo/fields.py [hunk 5]
```diff
@@ -1145,31 +1161,26 @@ def determine_domain(self, records, operator, value):
...
-                target = records - protected
-            elif path and env.in_onchange:
-                target = (env.cache.get_records(target, field) - protected).filtered(
-                    lambda rec: rec if path == 'id' else rec._mapped_cache(path) & records
-                )
+            if path == 'id' and field.model_name == record._name:
+                target = record
             else:
-                target = env.cache.get_records(target, field) - protected
+                target = env.cache.get_records(env[field.model_name], field)
```
Comment (rco-odoo) [inline odoo/fields.py]: Or only new records?

File: odoo/fields.py [hunk 6]
```diff
@@ -2346,12 +2362,12 @@ def convert_to_onchange(self, value, record, names):
 
         result = [(5,)]
         for record in value:
-            if not record.id:
+            if not record.id and not record._origin:
```
Comment (rco-odoo) [inline odoo/fields.py]: `convert_to_onchange` is actually deprecated for x2many fields...

File: odoo/models.py [hunk 7]
```diff
@@ -4963,31 +4981,45 @@ def update(self, values):
...
         return record
 
+    @property
+    def _origin(self):
+        """ Return the actual records corresponding to ``self``. """
+        return self.browse([
+            id_ or id_.origin
+            for id_ in self._ids
+            if id_ or id_.origin
+        ])
```
Comment (rco-odoo) [inline odoo/models.py]: Adapt prefetching after merging #32905.

File: odoo/models.py [hunk 8]
```diff
@@ -5488,40 +5522,42 @@ def diff(self, other):
...
+                        line_ids.update(cmd[2])
                 lines = self.browse()[name].browse(line_ids)
                 lines.read(list(subnames), load='_classic_write')
+                # copy values from lines to new lines
+                for line in lines:
+                    new_line = line.new(origin=line)
+                    new_line._cache.update({
+                        subname: line._cache[subname]
+                        for subname in subnames
+                    })
```
Comment (rco-odoo) [inline odoo/models.py]: This code is no longer useful if `new_line._origin` can prefetch all `lines`.

File: odoo/fields.py [hunk 9]
```diff
@@ -2276,31 +2287,36 @@ def process(ids):
...
+                browse = comodel.browse
+            # determine the value ids
             ids = OrderedSet(record[self.name].ids if record.id else ())
             # modify ids with the commands
             for command in value:
                 if isinstance(command, (tuple, list)):
                     if command[0] == 0:
-                        ids.add(comodel.new(command[2], command[1]).id)
+                        line = comodel.new(command[2], ref=command[1])
+                        ids.add(line.id)
```
Comment (rco-odoo) [inline odoo/fields.py]: What about doing `create` if `record` is real, and `new` otherwise?

File: odoo/models.py (hunk 8 above)
Comment (rco-odoo) [reply odoo/models.py]: It is actually necessary for computed fields that are not given by the client.

**Comment by celm1990** [comment]
@rco-odoo this commit must be solve issue #21510 and #21511 ???

File: addons/product/tests/test_product_attribute_value_config.py [hunk 10]
```diff
@@ -111,7 +111,7 @@ def _add_hdd_attribute_line(self):
         self.computer_hdd_attribute_lines.product_template_value_ids[2].price_extra = 8
 
     def _add_ram_exclude_for(self):
-        self._get_product_value_id(self.computer_ram_attribute_lines, self.ram_16).update({
+        self._get_product_value_id(self.computer_ram_attribute_lines, self.ram_16).write({
```
Comment (rco-odoo) [inline addons/product/tests/test_product_attribute_value_config.py]: necessary?

**Comment by rco-odoo** [comment]
@robodoo r+

**Comment by rco-odoo** [comment]
@robodoo rebase-ff r+

**Comment by rco-odoo** [comment]
@robodoo retry

**Comment by Yenthe666** [comment]
🤤 nice. Any insights on perf?

**Pull request closed without merging**